xline/src/storage/index.rs

```diff
@@ -185,89 +173,82 @@ impl IndexOperate for Index {
+                        .collect()
+                });
                 let keys = if pairs.is_empty() {
                     vec![]
                 } else {
                     vec![key.to_vec()]
                 };
                 (pairs, keys)
             }
-            RangeType::AllKeys => inner
-                .index
-                .iter_mut()
-                .zip(0..)
-                .filter_map(|((k, revs), i)| {
-                    Self::gen_del_revision(revs, revision, sub_revision.overflow_add(i))
-                        .map(|pair| (pair, k.clone()))
-                })
-                .unzip(),
-            RangeType::Range => inner
-                .index
-                .range_mut(KeyRange::new(key, range_end))
-                .zip(0..)
-                .filter_map(|((k, revs), i)| {
-                    Self::gen_del_revision(revs, revision, sub_revision.overflow_add(i))
-                        .map(|pair| (pair, k.clone()))
-                })
-                .unzip(),
-        };
-        if !keys.is_empty() {
-            assert!(
+            RangeType::AllKeys => self.inner.map_write(|mut inner| {
                 inner
-                    .unavailable_cache
-                    .insert(revision, keys.clone())
-                    .is_none(),
-                "revision {revision} is already in the unavailable cache",
-            );
-        }
+                    .index
+                    .iter_mut()
+                    .zip(0..)
+                    .filter_map(|((k, revs), i)| {
+                        Self::gen_del_revision(revs, revision, sub_revision.overflow_add(i))
+                            .map(|pair| (pair, k.clone()))
+                    })
+                    .unzip()
+            }),
+            RangeType::Range => self.inner.map_write(|mut inner| {
+                inner
+                    .index
+                    .range_mut(KeyRange::new(key, range_end))
+                    .zip(0..)
+                    .filter_map(|((k, revs), i)| {
+                        Self::gen_del_revision(revs, revision, sub_revision.overflow_add(i))
+                            .map(|pair| (pair, k.clone()))
+                    })
+                    .unzip()
+            }),
+        };
         (pairs, keys)
     }
 
-    fn insert_or_update_revision(
-        &self,
-        key: &[u8],
-        revision: i64,
-        sub_revision: i64,
-    ) -> KeyRevision {
-        let mut inner = self.inner.lock();
-        let new_rev = if let Some(revisions) = inner.index.get_mut(key) {
-            if let Some(rev) = revisions.last() {
-                let new_rev = if rev.is_deleted() {
-                    KeyRevision::new(revision, 1, revision, sub_revision)
+    fn insert(&self, key_revisions: Vec<(Vec<u8>, KeyRevision)>) {
+        self.inner.map_write(|mut inner| {
+            for (key, revision) in key_revisions {
+                if let Some(revisions) = inner.index.get_mut(&key) {
```

Comment:
here we could use `map.entry().or_insert()` syntax

Reply:
Done